src/eth1spec/spec.py

```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
     )
 
 
+def compute_header_hash(header: Header) -> Hash32:
+    """
+    Computes the hash of a block header.
+
+    Parameters
+    ----------
+    header : `eth1spec.eth_types.Header`
+        Header of interest.
+
+    Returns
+    -------
+    hash : `eth1spec.eth_types.Hash32`
+        Hash of the header.
+    """
+    return crypto.keccak256(
+        rlp.encode(
+            (
+                header.parent_hash,
+                header.ommers_hash,
+                header.coinbase,
+                header.state_root,
+                header.transactions_root,
+                header.receipt_root,
+                header.bloom,
+                header.difficulty,
+                header.number,
+                header.gas_limit,
+                header.gas_used,
+                header.timestamp,
+                header.extra_data,
+                header.mix_digest,
+                header.nonce,
+            )
+        )
+    )
+
+
+def get_block_header_by_hash(hash: Hash32, chain: BlockChain) -> Header:
+    """
+    Fetches the block header with the corresponding hash.
+
+    Parameters
+    ----------
+    hash : `eth1spec.eth_types.Hash32`
+        Hash of the header of interest.
+
+    chain : `eth1spec.eth_types.BlockChain`
+        History and current state.
+
+    Returns
+    -------
+    Header : `eth1spec.eth_types.Header`
+        Block header found by its hash.
+    """
+    for block in chain.blocks:
+        if compute_header_hash(block.header) == hash:
+            return block.header
+    else:
+        raise ValueError(f"Could not find header with hash={hash.hex()}")
+
+
+def validate_proof_of_work(header: Header) -> bool:
+    """
+    Validates the Proof of Work constraints
+
+    Parameters
+    ----------
+    header : `eth1spec.eth_types.Uint`
+        header of interest
+
+    Returns
+    -------
+    valid : `bool`
+        True if Proof of Work constraints are satisfied, False otherwise
+    """
+    # TODO: Implement this method once proof of work
+    #  algorithm is implemented
+    #  https://github.com/quilt/eth1.0-specs/issues/43
+    raise NotImplementedError
+
+
+def validate_gas_limit(gas_limit: Uint, parent_gas_limit: Uint) -> bool:
+    """
+    Validates the gas limit for a block
+
+    Parameters
+    ----------
+    gas_limit : `eth1spec.eth_types.Uint`
+        Gas limit to validate.
+
+    parent_gas_limit : `eth1spec.eth_types.Uint`
+        Gas limit of the parent block.
+
+    Returns
+    -------
+    valid : `bool`
+        True if gas limit constraints are satisfied, False otherwise
+    """
+    max_adjustment_delta = parent_gas_limit // GAS_LIMIT_ADJUSTMENT_FACTOR
+    if gas_limit >= parent_gas_limit + max_adjustment_delta:
+        return False
+    if gas_limit <= parent_gas_limit - max_adjustment_delta:
+        return False
+    if gas_limit < GAS_LIMIT_MINIMUM:
+        return False
```

Comment:
In `verify_header`, you do a series of `assert` statements followed by a `return True`, but here you use a series of `if` which can return `False`.

Is there any reason for the different styles?

Reply:
I would suggest this as a straw-man starting point for a convention for this codebase:

```python
def check_thing(...) -> bool:
    ...

def validate_thing(...) -> None:
    assert ...

is_thing_valid = check_thing(...)
if check_thing(...):
   ...
assert check_thing(...)

validate_thing(...)  # blows out an AssertionError or passes along quietly.
```

As an implementation detail: you can compose these like this:

```python
def check_thing(...) -> bool:
    ...  # actual implementation here
    
def validate_thing(...) -> None:
    assert check_thing(...)
```

This type of composition is "convenient" but it also loses any exception context since all you get is a `True/False`.  This can be inverted as:

```python
def validate_thing(...) -> None:
    assert condition

def check_thing(...) -> bool:
    try:
        validate_thing(...)
    except AssertionError:
        return True
    else:
        return False
```

This preserves detailed error information while still avoiding the need to duplicate logic.